R/utils.R

```diff
@@ -230,3 +230,43 @@ teal_data_datanames <- function(data) {
     ls(teal.code::get_env(data), all.names = TRUE)
   }
 }
+
+#' Function for validating the title parameter of `teal::init`
+#'
+#' Checks if the input of the title from `teal::init` will create a valid title and favicon tag.
+#' @param shiny_tag (`shiny.tag`)
+#' @keywords internal
+validate_app_title_tag <- function(shiny_tag) {
+  checkmate::assert_class(shiny_tag, "shiny.tag")
+  checkmate::assert_true(shiny_tag$name == "head")
+  child_tags <- sapply(shiny_tag$children, function(child) child$name)
+  checkmate::assert_true(
+    all(
+      c("title", "link") %in% child_tags
+    )
+  )
```

Comment:
```suggestion
child_names <- vapply(shiny_tag$children, `[[`, character(1L), "name")
checkmate::assert_set_equal(child_names, c("title", "link"), .var.name = "child tags")
```

Reply:
It might be a little restrictive. Technically, only the presence of these two is needed. A user can pass additional elements to this head tag. I can't think of any use case for this. But, I don't see any problem with keeping this open. So I would prefer to allow the users to pass additional tags along with `title` and `link`. We just need to check for the presence of these two. I agree with the checkmate simplification.

Reply:
I made it restrictive so that _only_ a tag generated by `build_app_title` can get through. Was that not the purpose of the validation function? Again, if you want looser criteria, why not go with just adding a class?

Reply:
Right now, there is flexibility to provide just an HTML tag as input without the need to use the helper function, which is more flexible than creating yet another custom class. We already have a lot of those and I don't feel that we benefit from this.
We indeed want the tags generated from the `build_app_title` to work. But, essentially it has to be an optional helper function. Hence, we should be supporting all the valid title tags generated by the user.

Reply:
Ok, so if you want `build_app_title` to be optional, you actually open the `title` argument significantly. Previously it accepted only character strings, now it accepts `shiny.tag`s. My original question is: are you sure you want to allow the user to pass _any_ `shiny.tag`, to which you answer seems to yes, as long as it can stick a favicon in the title. Are you sure you want to give the user this much wiggle room? _Any_ tag?

Conversely, if you add a special class and restrict `title` to that class _or character_, you give the user one more, very specific option. And they can still pass a bare string and you function will intercept it.

Reply:
From an app developer's point of view, they are free to pass any shiny tag in the header and footer. A valid shiny tag also includes a character.
So, I thought it's best to follow the same input types for the title too. Think about it like we have a validation of tag in place and a helper function to help create this tag.
<img width="954" alt="Screenshot 2023-12-24 at 10 54 50 AM" src="https://github.com/insightsengineering/teal/assets/49812166/93a4a1ca-3a0a-4b94-8ef1-b0f2e9f3bb0e">